### ISSUE TITLE
Rollout restarts of `ztunnel` node proxy result in proxy failures

### ISSUE THREAD

**Issue by bleggett** [istio/istio#43991]
### Bug Description

This is a weird one.

tl;dr is restarting ztunnel on a node with a rollout pattern breaks the mesh, whereas restarting ztunnel on a node with a "terminate-and-replace" pattern does not - and CNI is AFAICT not the culprit.

### Repro

1. Deploy Istio with ambient enabled.
2. Manually recycle all `ztunnel` pods on all nodes with e.g. `kubectl delete pod -n istio-system ztunnel-xyz`, and wait for Kube to reschedule them.
3. Run an ambient integration test, such as:
```sh
go test -tags=integ ./tests/integration/ambient/... --istio.test.ambient  --istio.test.ci -run "TestPodIP" -p 1 --istio.test.nocleanup  --log_output_level=all:debug  --istio.test.kube.helm.values="global.logging.level=all:debug" -v
```
4. Note that the test should *pass*
5. Recycle all ztunnel pods with a _rollout_ via e.g. `kubectl rollout restart ds/ztunnel -n istio-system`, and wait for the new pods to become ready.
6. Run the same test again and note that it begins to *fail, and from then on will continue to fail.* Not _all_ of them will fail (some of the waypoint/uncaptured tests work fine).
7. Note that manually restarting `ztunnel` pods without using `rollout` (e.g., restarting them with `delete pod`) will cause the tests to begin passing again.

Basically, any form of `ztunnel` restart that involves a _rollout_, where a new ztunnel pod is brought up alongside the old one on the same node, both of them are briefly "ready", and _then_ the old one is terminated breaks the mesh/tests.

Any form of `ztunnel` restart that involves a _recycle_, where the current ztunnel pod on the node is terminated _before_ the replacement comes up, works fine/"fixes" the mesh/tests.

I was convinced this was an `istio-cni` bug/race condition involving either
1. Not picking the ready ztunnel pod with the newest creation date as the "right" one, resulting in traffic getting redirected to an old/bad ztunnel IP.
2. CNI not properly cleaning up/recreating rules when transitioning redirection to the new ztunnel pod.

After debugging I'm actually pretty convinced that the CNI is behaving correctly - it _always_ selects the correct ztunnel pod with either mode of ztunnel start, and properly recreates all iptables, rules, tunnels, etc to point to the correct ztunnel pod. I can find no glitch or race condition there - the CNI behavior is the same between both restart methods.

In addition, I actually see `ztunnel` getting the traffic and making connections to upstream services even when the tests are broken.

Which is leading me to suspect this might be a bug in the control plane - some state somewhere is bad because in the `rollout` method, for a brief/temporary time there are two active ztunnel pods on a single node - even though in all cases CNI is only ever redirecting traffic to one or the other of them (or none of them, if neither is ready).

### Version

```prose
Current `master` ztunnel and `istio` as of 2023-03-17:


client version: 1.18-dev
control plane version: 1.18-dev
data plane version: 1.18-dev (20 proxies)
```
```


### Additional Information

_No response_

**Comment by linsun** [comment]
subscribe

cc @howardjohn if you have any ideas

**Comment by howardjohn** [comment]
funny, just saw the same issue at the exact same time. this is definitely a regression, it used to work.

**Comment by kdorosh** [comment]
> funny, just saw the same issue at the exact same time. this is definitely a regression, it used to work.

i can confirm this used to work in 76c45169acca08dad4599171265a7ae0dab2ea40 but broke when we updated to 2ffc98da4c8d9d8cda3c9164aad2bf4183396d1e . something in between those commits has the regression

**Comment by bleggett** [comment]
Oh, I think the problem is I'm an idiot. I see the problem. Fix incoming: https://github.com/istio/istio/commit/8d522cb39d3bda228f907c52e2bda37a6adbfc56#diff-6d7d78e80e2b7b324e6e8d26cf5f7a3f4b66d1bc9bdecd52850d0747ee97d019R129

This obviously won't work with rollouts as is.

**Comment by bleggett** [comment]
I'm still suspicious it could be something to do with CNI, but going back over the commits in those ranges doesn't throw up anything that would cause this - we don't restart CNI, just ztunnel.

**Comment by bleggett** [comment]
picking this back up.

**Comment by bleggett** [comment]
Closing this, cannot repro with latest `master`
